### ISSUE TITLE
Consumer group offsets and lag per partition in ok state

### ISSUE THREAD

**Issue by drax68** [linkedin/Burrow#22]
Hi, please add rendering of per-partition consumer group offsets and lags when overall group status is OK.
We need this to monitor if any message in any partition left unconsumed.

We used bin/kafka-run-class.sh kafka.tools.ConsumerOffsetChecker for this purpose before, but it can't handle offsets that's not stored in zookeeper.

**Comment by toddpalino** [comment]
It is possible to retrieve this information using the HTTP endpoint - https://github.com/linkedin/Burrow/wiki/http-request-consumer-topic-detail

Does that work for you?

**Comment by drax68** [comment]
Unfortunately not, we need lag as well (updated description). All info, that shown when consumer group has non-ok status.

**Comment by pparth** [comment]
+1

**Comment by pparth** [comment]
+1. I think the partitions lag data are very important to be always available when requesting the Status. This will help better with tweaking the evaluation algorithm parameters to the specifics of each system.

**Comment by toddpalino** [comment]
I agree that this information is very useful to have. We use information like this for our current lag checking (through a completely different system - it's all threshold based, which is what prompted me to create Burrow).

I had somewhat reserved an HTTP endpoint, /v2/kafka/(cluster)/offsets for performing requests like this. My intention was to implement this similar to how I have it internally on another app - it is a POST request, and in the body you can specify a list of topics and/or a list of consumers. If you just specify a topic, you get the broker head and tail offsets for that topic. If you just specify a consumer, you get the consumer and broker offsets for all topics consumed by the consumer. If you specify a consumer and a topic, you get the offsets for just that topic for both the broker and consumer.

Another thing we can do is provide a separate endpoint on the consumer (/v2/kafka/(cluster)/consumer/(group)/lag or something like that) which essentially dumps the information that Burrow is holding about the consumer. That is, for the number of intervals configured, the timestamp of the offset commit, the offset, and the lag at that time. I think that's more along the lines of what you're looking for.

What I'd like to avoid is loading too much information into the "status" endpoint. Right now it provides the evaluation and any errors present. I think if it also dumped the lag information regardless of status, it's too much information. Especially for wildcard consumers like a mirror maker.

**Comment by pparth** [comment]
The solution of providing a separate endpoint on the consumer, should do the trick. When can we expect that?

For the status endpoint, it would be useful to add information about the partition with the maximum lag, when status is OK.

**Comment by toddpalino** [comment]
That sounds like a good idea as well.

As far as the timeframe goes, none of this is incredibly complicated. I'm tied up with some presentation work this week, but I should be able to find time early next week to work on it.

**Comment by toddpalino** [comment]
Thanks for bearing with me on this. I've added both features to Burrow now. The partition with the maximum ending lag is always returned in a status request. There is also a new HTTP endpoint for consumer groups `/lag`. This is exactly the same as the `/status` endpoint, except it returns every partition object regardless of the evaluated state.

I've also tagged and released this as version 0.1.0, as we needed to start having release versions.
